### ISSUE TITLE
Apple silicon builds

### ISSUE THREAD

**Issue by AshCripps** [nodejs/build#2474]
refs: https://github.com/nodejs/TSC/issues/886#issuecomment-722366519

We need to kickstart the discussion again on how we plan to support the new Mac hardware which is due to be announced on [Nov 10th](https://www.independent.co.uk/life-style/gadgets-and-tech/apple-event-macbook-iphone-ipad-b1563449.html). 

We last had the discussion in the [build meeting in july](https://github.com/nodejs/build/blob/master/doc/meetings/build-wg-meeting-2020-07-29.md#nodejsbuild) where we had some discussion as to whether to do 2 binaries or to go for fat binaries.

 We also need to discuss sourcing hardware - We currently have two DTKs kindly donated by macstadium but no release machine as of yet and these DTKs will go back to apple at some point.

To Do:
- [x] Add 11.0 to versionselector to ensure we build only on master
- [ ] Enable the macos-arm job in normal test regressions
- [ ] Get the DTKs back into CI for testing 
- [ ] Test compiling and running node natively on the mac arm hardware
- [ ] Test if a current release runs on the new hardware through the translation layer
- [ ] Figure out a source of hardware to add to our CI for when our DTKs go back to apple
- [ ] Note down size difference in binaries between: intel, arm and fat.

**Comment by mhdawson** [comment]
I think we'd agree to do fat binaries as they seemed preferable for a number of reasons.

In terms of hardware longer term, we should see if macstadium has a longer term answer for that now as having them provide the hardware is the easiest path for us.

**Comment by rvagg** [comment]
+1 to fat (again, I think? I don't recall). We used to do fat for x86+x64 until we ripped that out for io.js (and then Node 4). There's still echoes of the code to do it in nodejs/node Makefile and tools/ and it's certainly in the git history.

For hardware, we really should have redundancy here too. Same approach as we have for x64. Rely on macstadium (assuming they'll let us use official hardware for this) but get our own and ask nearForm if they'll kindly host them for us as backups.

**Comment by gengjiawen** [comment]
We have PRs need to run on silicon too: https://github.com/nodejs/node/pull/35986.

**Comment by mhdawson** [comment]
@rvagg redundancy makes sense.  I don't think we can order ARM mac mini H/W yet, correct me if I'm wrong on that and I can start the discussion with the Foundation.

**Comment by rvagg** [comment]
No, we're really just waiting to see Apple's announcement tomorrow. Rumours are suggesting laptops, which is probably not going to be helpful. But 🤞 we may get mini's out of this in which case we should jump straight on it.

**Comment by AshCripps** [comment]
Yeah the event looks to be the announcement of just 2 laptops according to rumours at least - we might be another year or so away from iMacs and mac minis.

**Comment by AshCripps** [comment]
Turns out they did announce new mac minis so it might be worth speaking with the guys at macstadium again to see their timeline.

**Comment by rvagg** [comment]
@mhdawson can you start the discussion with the Foundation about getting a couple of minis? And we'll need to make sure we have somewhere to put them, I guess we need to probe nearForm on this unless we have other options.

**Comment by AshCripps** [comment]
just as an FYI I am working to get the DTKs into our regular regression runs to get some testing at least on the platform - I managed to brick one updating to the latest big sur but macstadium have swapped it out for another one. Im waiting to hear back if its safe to update the second one before I carry on.

**Comment by mhdawson** [comment]
@rvagg last time we also asked for virtualization software so that we could have a 2 vms on each mac mini so that 1 for test and one for release would be good enough.  Do you know if the same software will work on the arm machines?

We'll also need to agree on the specs. From https://www.apple.com/ca/shop/buy-mac/mac-mini there only seem to be 2 choices, one with the only difference being the SSD size.

**Comment by AshCripps** [comment]
Key difference this time is the minis are 8 core instead of 6 core so we _may_ be able to get away with running 3 VMs on one mini, The next issue is VMware fusion which I cant seem to find any more mention besides "soon" or "we are working on it" for M1 support

**Comment by nschonni** [comment]
I think there is also supposed to be support for 16GB or ram https://www.anandtech.com/show/16235/apple-intros-first-three-apple-silicon-macs-late-2020-macbook-air-13inch-macbook-pro-mac-mini but it doesn't look like it's on the shop either

**Comment by rvagg** [comment]
Oh, right yeah, the VM problem. I bet VMWare aren't treating this as a top priority. I wonder if we need to shop around for that. Maybe Orka is the only game in town for now unless we're happy to go bare metal for these?

**Comment by nschonni** [comment]
Looks like the license was updated around virtualisation https://blog.macstadium.com/blog/developers-big-sur-and-vindication

**Comment by rvagg** [comment]
So .. here's the key piece that someone needs to figure out - how do we get our toolchain to compile arm64 macos binaries on x64?

Here's how we used to make fat binaries: https://github.com/nodejs/node/blob/archived-io.js-v0.10/Makefile#L291-L301

i.e. we could do a `--dest-cpu=ia32` build and a `--dest-cpu=x64` build and smush the resulting `node` binaries together with `lipo`. This second piece is still the right way as far as I understand, the problem is the `--dest-cpu=arm64` bit.

Here's some things I've tinkered with so far on my x64 mac:

* Installed Xcode 12 beta which I believe is the only way to get this to work (I could be corrected on this, the fact that it's still beta has me confused about this point)
* Manually added `'SDKROOT': '/Library/Developer/CommandLineTools/SDKs/MacOSX11.0.sdk/'` to my common.gypi in the `'target_arch=="arm64"'` section for the `"mac"` block (without this it keeps on defaulting to my system SDK, I assume if we built on the latest OS this would default to the right directory).
* `./configure --dest-cpu=arm64` and `make`
* Added `--without-intl` to `./configure` because `icupkg` breaks the build - this is an expected part of cross compiling, we probably need a system ICU installed to make this work (although this isn't a great option for release machines where we want to push the ICU version with the source so we might need to figure out how to first compile it for x64 then use that binary to run against the arm64 build).

But my latest failure is V8 related: `OSError: [Errno 86] Bad CPU type in executable: '/Users/rvagg/git/iojs/io.js/out/Release/bytecode_builtins_list_generator'`.

I haven't got any further than that. Also, please don't count on me to solve all of these problems, I don't really have time to be doing any of this and was really just checking to see how far off we might be -- this all suggests that the answer is "quite a way off, and it will require coordinated effort across subsystems to make this happy".

**Comment by rvagg** [comment]
It seems `CC_target` and `CC_host` can be (ab)used here to get further: `CC_host="ccache cc -arch x86_64" CXX_host="ccache c++ -arch x86_64" CC_target="ccache cc -arch arm64" CXX_target="ccache c++ -arch arm64" ./configure --dest-cpu=arm64`, although that gets me to asm errors in `out/Release/obj.host/v8_base_without_compiler/deps/v8/src/heap/base/asm/x64/push_registers_asm.o.d.raw` during the x64 phase. But that's very late in the build. It could be because this approach results in a double `-arch` being supplied and maybe the later one is overriding the former.

**Comment by AshCripps** [comment]
https://developer.apple.com/documentation/xcode/building_a_universal_macos_binary Looks like Apple says to use `lipo` too but I dont know enough about the node source code to say if we have those conditional macros in place.

**Comment by rvagg** [comment]
Yeah, it's the two `-arch`'s that are appearing in the compile commands with the later one overriding the first. So if I remove the two lines that add the `-arch` from `tools/gyp/pylib/gyp/xcode_emulation.py` and rely on the ones supplied from my `CC_host` and `CC_target`:

`CC_host="ccache cc -arch x86_64" CXX_host="ccache c++ -arch x86_64" CC_target="ccache cc -arch arm64" CXX_target="ccache c++ -arch arm64" CC="ccache cc -arch arm64" CXX="ccache c++ -arch arm64" ./configure --dest-cpu=arm64`

Then I get what appears to be an arm64 binary:

```
$ file node
node: Mach-O 64-bit executable arm64
```

That's probably not super useful other than confirming that it is possible to get a cross-compiled binary out of our toolchain, if lightly hacked. It doesn't suggest that path to getting this working properly though. At this stage I'm suspecting that something about the CFLAGS changed in GYP along the way that messed this up. I don't think xcode_emulation.py has really changed much in all that time, but make.py has, including some of the CFLAGS logic (including the _host and _target stuff).

FTR the diff that gets me here is:

```diff
diff --git a/common.gypi b/common.gypi
index 4745bb5ac7..1e00c64f0f 100644
--- a/common.gypi
+++ b/common.gypi
@@ -510,6 +510,7 @@
           }],
           ['target_arch=="arm64"', {
             'xcode_settings': {
+              'SDKROOT': '/Library/Developer/CommandLineTools/SDKs/MacOSX11.0.sdk/',
               'ARCHS': ['arm64'],
               'OTHER_LDFLAGS!': [
                 '-Wl,-no_pie',
diff --git a/tools/gyp/pylib/gyp/xcode_emulation.py b/tools/gyp/pylib/gyp/xcode_emulation.py
index 8af2b39f9a..fa1bd07ab7 100644
--- a/tools/gyp/pylib/gyp/xcode_emulation.py
+++ b/tools/gyp/pylib/gyp/xcode_emulation.py
@@ -663,7 +663,7 @@ class XcodeSettings(object):
             # TODO: Supporting fat binaries will be annoying.
             self._WarnUnimplemented("ARCHS")
             archs = ["i386"]
-        cflags.append("-arch " + archs[0])
+        # cflags.append("-arch " + archs[0])

         if archs[0] in ("i386", "x86_64"):
             if self._Test("GCC_ENABLE_SSE3_EXTENSIONS", "YES", default="NO"):
@@ -947,7 +947,7 @@ class XcodeSettings(object):
             # TODO: Supporting fat binaries will be annoying.
             self._WarnUnimplemented("ARCHS")
             archs = ["i386"]
-        ldflags.append("-arch " + archs[0])
+        # ldflags.append("-arch " + archs[0])

         # Xcode adds the product directory by default.
         # Rewrite -L. to -L./ to work around http://www.openradar.me/25313838
```

**Comment by rvagg** [comment]
Oh, so here's the most plausible explanation as to why it used to work for our fat binary builds but now doesn't - the i386 build, simply using `--dest-cpu=ia32` built binaries (including the build-time tooling) that actually ran on the compile host, because the compile host could run i386 binaries as well as x64. So `-arch i386` got passed through to everything and it didn't matter, the resulting binaries could run regardless of whether it was for target or host. But now we need differentiation for cross-compiling, like we do for the ARMv7 cross compile, where there's a separation of these arch parameters between `_host` and `_target` so we get build-time tooling in the host's architecture and get the proper `_target` architecture in the resulting binary.

So it's this hard-wired `-arch` in xcode_emulation.py that's our main problem. Either we need it to properly switch between `_host` and `_target` (the elegant way) or we get to override it with the custom environment variables like I've done above (the hacky way).

@nodejs/gyp this is going to have to involve you to some degree, I don't know how familiar you are with xcode_emulation.py.

**Comment by rvagg** [comment]
A proposal for shoe-horning this in https://github.com/nodejs/gyp-next/pull/78

**Comment by AshCripps** [comment]
we held a meeting today and come up with the following next steps as a starter:

- [ ] Get the DTKs back into CI for testing 
- [ ] Test compiling and running node natively on the mac arm hardware
- [ ] Test if a current release runs on the new hardware through the translation layer
- [ ] Figure out a source of hardware to add to our CI for when our DTKs go back to apple
- [ ] Note down size difference in binaries between: intel, arm and fat.

as a note we need to figure out what breaking changes we might introduce with the apple silicon support and plan around targeting a specific release (e.g. 16.x)

Also need to ensure we get a green CI run on the new hardware as there have been some untested (by us) prs around apple silicon land on master recently.

**Comment by STRML** [comment]
Just want to confirm since I got one in the mail - no compatibility problems with node, it just works - tested with a large/complex project-  and even runs very well compared to Intel chips. So new owners should not be complaining this week.

The work is still worth doing given the ~30% performance boost it unlocks.

**Comment by aduh95** [comment]
FYI: https://github.com/nodejs/node/issues/36160

**Comment by akx** [comment]
Doesn't seem to be quite 100% problem-free: https://github.com/facebook/create-react-app/issues/10090

**Comment by eugene1g** [comment]
@akx some WASM-related things for Apple silicon have been fixed by https://github.com/nodejs/node/pull/35986 so that issue is now resolved in `master` :)

**Comment by richiksc** [comment]
I'm confused about the status of Apple Silicon support on node. I've seen many comments on issues that point to Apple Silicon being natively supported on `master` with nodejs/node#34238 and nodejs/node#35986 having landed in master for a while now  - have these changes made their way to a stable release of v15.x yet? 15.3.0 was released after both those PRs were merged. Does the `.pkg` installer downloaded from nodejs.org support darwin-arm64 natively?

**Comment by eugene1g** [comment]
@richiksc I can run Node natively (arm64 binary) on Apple Silicon. The most recent release 15.3 resolve last issues I noticed (related to WASM), and I can run large & complex codebases (both frontend and backend) with no issues whatsoever.

However, Apple Silicon is not yet _officially supported_ by the Node foundation. Official support is a big claim, and understandably the foundation is slow to take on that responsibility. There are no blessed downloadable binaries for arm64. I'm sure this will change very soon, but today you have to build Node either yourself or via Homebrew etc to get a native darwin-arm64 binary.

**Comment by paolotremadio** [comment]
No pre-built binaries yet. I’m running node 15.3.0 on M1 with no issues. The easiest way to install it is via nvm. It will try to download a binary (which it will not find) so it will compile one. 

All I had to do was `nvm install v15` and that was it. I had to install Xcode build tools when prompted and wait for a few minutes for node to build. No homebrew needed. 

It runs. It’s quick. No compatibility issues. There’s a Node meeting on the 1st of Dec and the support is on the agenda. We will know more in the next few weeks.

**Comment by wreiske** [comment]
If the nodejs project needs an M1 mac or mac mini for building and testing, please get in contact with me. I'd be happy to buy some mac minis or airs for the team to test with.

**Comment by Dids** [comment]
> No pre-built binaries yet. I’m running node 15.3.0 on M1 with no issues. The easiest way to install it is via nvm. It will try to download a binary (which it will not find) so it will compile one.
> 
> All I had to do was `nvm install v15` and that was it. I had to install Xcode build tools when prompted and wait for a few minutes for node to build. No homebrew needed.
> 
> It runs. It’s quick. No compatibility issues. There’s a Node meeting on the 1st of Dec and the support is on the agenda. We will know more in the next few weeks.

One thing to note with this is that you need to make sure you install/use `nvm` on the Terminal app, otherwise there's a chance it'll assume x86 for the architecture.

You should also be able to work around this by prefixing all commands with `arch arm64`. Additionally, you can run `arch` to see which architecture you're currently running on.

Disclaimer: I was using a 3rd party terminal app, which caused everything to run run through Rosetta, as it was emulating x86 for the entire terminal session.

**Comment by arthurlockman** [comment]
> > No pre-built binaries yet. I’m running node 15.3.0 on M1 with no issues. The easiest way to install it is via nvm. It will try to download a binary (which it will not find) so it will compile one.
> > All I had to do was `nvm install v15` and that was it. I had to install Xcode build tools when prompted and wait for a few minutes for node to build. No homebrew needed.
> > It runs. It’s quick. No compatibility issues. There’s a Node meeting on the 1st of Dec and the support is on the agenda. We will know more in the next few weeks.
> 
> One thing to note with this is that you need to make sure you install/use `nvm` on the Terminal app, otherwise there's a chance it'll assume x86 for the architecture.
> 
> You can run `arch` to see which architecture you're currently running on.
> 
> Disclaimer: I was using a 3rd party terminal app, which caused everything to run run through Rosetta, as it was emulating x86 for the entire terminal session.

I couldn’t actually get `nvm` to build a native version of Node even when running in a native terminal. It would build from source each time but produce an x64 binary.

Using the native version of Homebrew (installed in `/opt/homebrew`, just `brew install node`) I was able to produce an arm64 binary. Works a treat!

**Comment by HenkPoley** [comment]
Tip: The standard `nvm` install script uses uses `bash`. Which might be x86_64 on your system. Causing the whole chain to build under Rosetta 2.

```
% file `which bash`
/usr/local/bin/bash: Mach-O 64-bit executable x86_64
```

**Comment by AshCripps** [comment]
FYI. Ive gotten one DTK hooked up the CI and running and had to cross some hurdles. Firstly im running without ccache as that was failing to build from source from brew (I haven't bothered to build it fully from source yet because the build times seem to be fast enough for me currently). Ive also had to install a new java on the machine - the one we had was for x86_64 causing the jenkins job to build x64 binaries instead of arm64 so ive installed a aarch64 macos openJDK from azul as adopt dont have their available yet. 

Example binary:

```
bash-3.2$ ./out/Release/node -p process.arch
arm64
```

interesting when building x64 binaries I get no test failures but I get 2 for arm64 -    ` test.abort/test-signal-handler` and 
    `test.sequential/test-worker-prof` on https://ci.nodejs.org/job/node-test-commit-osx-arm/nodes=osx11/17/ so that needs some investigation.

Ive got some changes ive had to make to jenkins scripts + the jobs which I will upstream sometime this week once Ive ironed out some bugs to make it simpler to add the second DTK

**Comment by richardlau** [comment]
For `test.abort/test-signal-handler` I think we can just add `SIGTRAP` to the list of signals tested for in https://github.com/nodejs/node/blob/393a1feeee69c6fb1a66d2abb4ad070a49ee3850/test/abort/test-signal-handler.js#L20-L22

**Comment by AshCripps** [comment]
@richardlau thanks Ill try that out now

**Comment by BraydenGirard** [comment]
So is there any way to get Node 14 working on Apple Silicon? I have Node 15 working fine but some projects I am working with only support the LTS version of Node (Strapi).

**Comment by andykenward** [comment]
> So is there any way to get Node 14 working on Apple Silicon? I have Node 15 working fine but some projects I am working with only support the LTS version of Node (Strapi).

See https://github.com/nvm-sh/nvm/issues/2350#issuecomment-734132550 about running node versions via Rosetta 2.

**Comment by matinzd** [comment]
As of now we only have node v15 native support for darwin arm64?

**Comment by bdkjones** [comment]
Can anyone provide an update on the status of pre-built Universal binaries becoming available? The Apple Silicon thread on the TSC repo got locked. There are a few folks saying they were able to build node 15 from source if installed via Homebrew. Others have failed to build that way.

In short, it seems like there is no cohesive plan or timeline to get Universal Binaries shipped. (I'm sure SOMEONE has one, but it would be nice to get that stated publicly.)

**Comment by richiksc** [comment]
@bdkjones Homebrew now has official support for Apple Silicon, and there is a Node 15 pre-built bottle available.

**Comment by AshCripps** [comment]
We made a tentative plan at the start of December in a Working Group meeting of which notes can be found [here](https://github.com/nodejs/build/blob/master/doc/meetings/build-wg-meeting-2020-12-01.md#nodejsbuild).

TL:DR we are trying to target 16.x for the release of the fat binaries - currently you can build 15.x from source but you will not be able to build anything older than 15.x nor will you likely ever as prior release lines don't have the necessary V8 versions.

There is a checklist at the top of this issue of next steps - CI is currently blocked on a failing test tracked here -> https://github.com/nodejs/node/issues/36656

**Comment by gpetrov** [comment]
@AshCripps what is the reason to choose for fat binaries and should all native node modules also target fat binaries?

**Comment by andreialecu** [comment]
@gpetrov imo all native node modules should target fat binaries, yes. See https://github.com/fsevents/fsevents/issues/349 for a related discussion.

Also FWIW:

Apple recently broke nodeJS in the latest Release Candidate of macOS 11.2:
https://github.com/nodejs/node/issues/37061

**Comment by AshCripps** [comment]
The reason we chose fat binaries was for ease on us - having two separate binaries will add layers of complexity to our release process and also increase the strain on our download server by hosting another binary.

**Comment by mhdawson** [comment]
Adding to what @AshCripps mentioned, it also follows the pattern that work for the previous transition from Power to Intel.

**Comment by tholu** [comment]
`./configure` followed by `make DESTCPU=arm64 binary` worked fine for me when building binaries for v15.6.0 from source on 11.1 (20C69).

**Comment by bdkjones** [comment]
@tholu I tried your method using `15.10.0` downloaded directly from nodejs.org:

```
./configure
make DESTCPU=arm64 binary
```

I'm on an M1 Mac and I have Xcode 12.4 installed. I get the same error I got trying to build previously:

![Screen Shot 2021-02-24 at 01 33 45](https://user-images.githubusercontent.com/926566/108980572-a8635a00-7640-11eb-8864-36a17a14354d.png)

If this is building for everyone else, I find it hard to believe the makefile is incorrect? But libtool seems to think so?

**Comment by bdkjones** [comment]
@tholu Sigh. It's a known-issue with Node's build chain that goes back at least six years. For anyone else who's looking to build, the answer is here: https://github.com/nodejs/node/issues/2341

TL;DR: Node needs an exact, specific version of `libtool`.

**Comment by bdkjones** [comment]
@tholu did your arm64 build successfully past all tests? I have two failed tests when running `make test-only`

**Comment by tholu** [comment]
@bdkjones I'll try again later today and will let you know!

**Comment by bdkjones** [comment]
@tholu I got the same test failures building 15.10.0 on x86, so it doesn’t appear to be related to the platform. Don’t recall the exact failure now, but something about “parallel promisify executed too early.”

**Comment by nschonni** [comment]
@tholu @bdkjones similar to comment on https://github.com/nodejs/TSC/issues/886#issuecomment-741943777, https://github.com/nodejs/help/ is the better place if you're trying to debug why it's not working on your machine vs the discussion around getting it running in CI to produce releases (here)

**Comment by bdkjones** [comment]
@nschonni Thanks. From the perspective of an outsider, I don't think Node has done a very good job explaining their timeline or their progress on this issue. As someone at nodejs/TSC#886 pointed out: there is no single source of truth for those of us trying to get Node running on `arm64`—you have to cobble together bits and pieces from several threads all over the place. The people with the expertise are in these threads, so it's natural to talk to them there.

If you want to avoid that in the future, when large-scale platform transitions occur, it would be a good idea to update the "releases" page at nodejs.org with a link that says something like, "Trying to build Node on that new M1 Mac? Here's what you need to know." Centralize the expertise and you won't be bothered by the unwashed idiot masses (me).

**Comment by AshCripps** [comment]
There is a pinned issue on nodejs/node which is as up to date as to our progress - https://github.com/nodejs/node/issues/37309

**Comment by kalwiggins** [comment]
anyone get node 10 to work on apple silicon? I haven't been able too

**Comment by architchandra** [comment]
Can't get Node `10.14` to install using `nvm` on an M1 Mac Air. Any workarounds?

**Comment by billinghamj** [comment]
I would say that is expected. Node 15 supports M1. For older versions, use Rosetta

**Comment by arcreigh** [comment]
All my M1 mini is in route, while I am not a developer I want to assist in any way that I can. Even if it means picking up a new skill / helping test compilation or running various tests.

**Comment by AshCripps** [comment]
Closing. Support for Apple silicon is 15.x and above, and the binaries will be released with the 16.x release next week

**Comment by alcuadrado** [comment]
Should this be working now? I'm using volta `1.0.4` and `volta install node@16` installs an x86_64 version.

**Comment by marckohlbrugge** [comment]
Are the Apple Silicon builds for 15.x still planned? I can see them for [16.x](https://nodejs.org/dist/latest-v16.x/) but not for [15.x](https://nodejs.org/dist/latest-v15.x/)

**Comment by targos** [comment]
No, and there are no more releases planned for v15.x

**Comment by 7rulnik** [comment]
@targos what about 14.17+?

**Comment by AshCripps** [comment]
@7rulnik 14.x releases of apple silicon are currently not planned but the support is in the code base if you wish to build from source.
